github/commitstatus.go

```diff
@@ -112,10 +112,8 @@ func (s CommitStatus) Add(sha, state, targetURL, description string) error {
 	req.Header.Set("Content-Type", "application/json")
 
 	var response httpResponse
-	timeToSleep := 0 * time.Second
 
 	for attempt := 1; attempt <= s.target.MaxRetries; attempt++ {
-		time.Sleep(timeToSleep)
```

Comment:
When I did my previous implementation there were two hidden marvels in this for loop.

1. avoid shadowing `:=` 
2. put `time.Sleep(timeToSleep)` at the top and not at the bottom.

At the time, I couldn't make it to work when I added `time.Sleep(timeToSleep)` at the end of the iteration like you are doing now. Unfortunately, I forgot what was the reason and now looking at your code I can't find any reason why it wouldn't work as proposed in this PR. 

I will go back a bit to the previous version of the code, to try to identify the reason. I just wanted to mention here to be careful a bit.

Reply:
I _think_ it was only to avoid a spurious sleep(0), although it would have been harmless.

Reply:
See also especially approach 2, where the tests are actually testing everything and should make it clear that the code is now correct.

Reply:
Ok. This can't occur with your code.

It was specific to that implementation at the time.
Here is the discussion: https://github.com/Pix4D/cogito/pull/118#discussion_r1173377510
and commit that moved it up: https://github.com/Pix4D/cogito/commit/c8f7ab2d1fbd2f8ea898ee590f2500d2b7d5fe71

Reply:
actually, wait... it's still important. If max attempt is reached you will sleep for `waitTime` but there will not be next attempt. in example, for 500 status codes.

Reply:
Ah. I think I understand what you mean. We are missing a test also for this case.

The more I touch this code, the more I think that an exponential back-off with a stop only on a time limit, no matter the number of attempts, would be simpler and so with less corner cases. We went with counting the attempts in an effort to "optimize" the number of requests towards GitHub since we knew when the reset would happen, but this caused the code to be fragile. Such optimization is not needed.

I will add the missing test and a fix for this behavior.

Reply:
Note also, that my preferred approach is option 2. So I will continue with comments there, and just "ignore" this PR.

Reply:
For sure. This approach is here to show how to evolve an approach to something better.